### ISSUE TITLE
Tree Shaking

### ISSUE THREAD

**Issue by jacobp100** [MikeMcl/decimal.js#162]
What are your thoughts on making all functions ES6 exports (like `add` etc), then providing a 'lite' constructor with no functions on the prototype? This would mean for people not calling the functions on the constructor, we can tree shake.

[My project](https://github.com/jacobp100/technicalc-calculator/blob/master/src/Decimal/Decimal.re) uses ReScript (/ReasonML), so I just write bindings to whatever the exports are.

In the app, the calculations are done in a web worker, so pretty much requires the whole of this library. Once calculated, the results are sent back to the client, but the only functionality I need from this library is `toString`. Having this library be tree-shakable would significantly reduce the size of my client file.

**Comment by MikeMcl** [comment]
Yes, that would be straightforward, but I would release it as a separate ES6 library (decimal.js will remain ES3 compatible).

I'll do it for £250 (bitcoin/paypal) for a delivery time of 5 weeks, more if you want it sooner.

**Comment by jacobp100** [comment]
Thanks for the offer! I can handle doing this myself - I just wanted to know if you were interested in having it in the library. I'll aim make a copy of this library with the fix in the project I mentioned. You're welcome to copy it back if you want.

**Comment by jacobp100** [comment]
My work is over at https://github.com/jacobp100/technicalc-calculator/blob/master/src/Decimal/_Decimal.js

Still in the process of getting all the tests to pass - but a few of the suites are passing already

**Comment by MikeMcl** [comment]
I'm committed to keeping this library ES3 compatible, but I'll take a look. It's a good idea.

**Comment by jacobp100** [comment]
Have you looked into adding build tools to this? It should be possible to write just one version (ES3 but with ES6 imports/exports), then produce an ES3-only bundle. That's something I could help with if you need!

**Comment by MikeMcl** [comment]
Yes, that has been considered, but at the moment I prefer to just include a separate ES module file, *decimal.mjs*.

**Comment by jakub791** [comment]
> I'm committed to keeping this library ES3 compatible, but I'll take a look. It's a good idea.

ES5 is supported everywhere now, including Internet Explorer which is like 10 years old. You do not need to support anything below that. If someone's browser does not support ES5 (or even ES6 now) then that shouldn't be our issuse.

**Comment by jakub791** [comment]
> Yes, that has been considered, but at the moment I prefer to just include a separate ES module file, _decimal.mjs_.

Why does it use ES3 compatible syntax? Modules were added in ES6.

**Comment by WuglyakBolgoink** [comment]
Any updates?

**Comment by MikeMcl** [comment]
@WuglyakBolgoink 

Yes, I have a bad cold and can't concentrate on anything.

**Comment by WuglyakBolgoink** [comment]
> @WuglyakBolgoink
> 
> Yes, I have a bad cold and can't concentrate on anything.

get well soon!!!

**Comment by MickL** [comment]
Hey guys! I saw there is a light version of decimal.js and was thinking that this wouldnt be needed at all if decimal.js would support tree shaking properly. This could be done by using individual imported functions which would also have the benifit of this functions to be better testable because they have clear inputs and outputs:

Current code:

```ts
x = new Decimal(255.5)
x.toExponential(5)                       // '2.55500e+2'
x.toFixed(5)                             // '255.50000'
x.toPrecision(5)
```

Possibly new syntax:

```ts
import { Decimal, toExponential, toFixed, toPrecision } from 'decimal.js';

x = new Decimal(255.5)
   .pipe(
      toExponential(5)                       // '2.55500e+2'
      toFixed(5)                             // '255.50000'
      toPrecision(5)
);
```

**Comment by josdejong** [comment]
an other option would be to change from methods into functions, like:

```js
import { Decimal, add, subtract } from 'decimal.js'

const a = new Decimal("0.1")
const b = new Decimal("0.2") 
const c = add(a, b)
const d = subtract(a, b)
```

**Comment by MickL** [comment]
Yeah thats exactly what I meant :) Except for optionally a `pipe()` function could be added to `Decimal `

**Comment by josdejong** [comment]
Ah, true, a pipe function can pipe those static functions like `add(a, b)`

**Comment by jacobp100** [comment]
@MickL I’d be sceptical of the pipe function for performance reasons. Not only do you need to allocate for the temporary function, if any arguments are floats, they’d also need to be boxed - so even more allocations. Direct function calls are much more efficient

**Comment by MikeMcl** [comment]
I have an enhanced tree-shakeable version of decimal.js but it's only 95% of the way there and I've been side-tracked for the last year, so it's not quite ready for publication.
